### PR TITLE
TT-7086 - feat(helm): Add hostNetwork option

### LINKED ISSUE #532: [TT-7086] Add hostNetwork option in helm chart
## Expected Behavior

Given our current ENI Configuration in Amazon EKS we need to run the Tyk Operator deployment with hostNetwork mode set to true.  I'd like to be able to specify this as a boolean toggle in the values file.

### Potential BDD Test Coverage

```gherkin
Feature: hostNetwork Mode
  As a developer
  I want to enable hostNetwork mode for the deployment
  So that I can install Tyk Operator in my EKS cluster

  Scenario: hostNetwork mode enabled
    Given that .Values.hostNetwork = true
    When I deploy Tyk Operator
    Then the associated pods run with hostNetwork mode enabled
```

## Current Behavior

The current helm chart provides no means to update the hostNetwork value so I am unable to configure this without editing the underlying template.  I'd like to bring this chart in as a helm dependency and not have to edit the template. 

## Possible Solution

A small block could be added [here](https://github.com/TykTechnologies/tyk-operator/blob/6f4c531a58486ef8786a6a4c6577354a2bfe89b2/helm/templates/all.yaml#L418) to set hostNetwork mode equal to .Values.hostNetwork mode or a default of false

```yaml
      hostNetwork:  {{ .Values.hostNetwork | default false }}
```

### PR DESCRIPTION
**Pull request by rdcwaldrop1**
## Description

Add the option to turn host networking on.  With the host networking option on it is likely users will also want to control port selection so as not to collide with other pods running with hostNetwork mode enabled.  To accomodate port selection the ports have been parameterized in the template as well.

## Related Issue

Resolves #532 - [TT-7086](https://github.com/TykTechnologies/tyk-operator/issues/532)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If PRing from your fork, don't come from your `master`!
- [x] Make sure you are making a pull request against our **`master` branch** (left side). Also, it would be best if you started *your change* off *our latest `master`*.
- [x] Make sure you are updating [CHANGELOG.md](https://github.com/TykTechnologies/tyk-operator/blob/master/CHANGELOG.md) based on your changes.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] If you've changed API models, please update CRDs.
  - [ ] `make manifests`
  - [ ] `make helm`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `gofmt -s -w .`
  - [ ] `go vet ./...`
  - [ ] `golangci-lint run`

[TT-7086]: https://tyktech.atlassian.net/browse/TT-7086?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by buraksekili** [comment]
@rdcwaldrop1 thank you for this great contribution. It looks great but we need to change the approach for this problem a bit. Let me briefly talk about how we generate helm manifests.

[`helm/templates/all`](https://github.com/TykTechnologies/tyk-operator/blob/c4b85072dce3a66eae6dce59e1d7eb20090f8a86/helm/templates/all.yaml) file is created automatically with [`make helm`](https://github.com/TykTechnologies/tyk-operator/blob/c4b85072dce3a66eae6dce59e1d7eb20090f8a86/Makefile#L65) command. Therefore, direct updates on the `helm/templates/all` file will be overwritten by the next make helm command, which will eventually override your contribution.

What we do for creating a helm manifest is to run `make helm` command. This command runs `kustomize` on `config/helm` and sends the output of `kustomize` to [`hack/helm/pre_helm.go`](https://github.com/TykTechnologies/tyk-operator/blob/cd71d5254803def0a80ff563794f0a3b200fcf74/hack/helm/pre_helm.go) file which replaces certain fields with their templating correspondence.

For example, in order to update [`helm/templates/all`](https://github.com/TykTechnologies/tyk-operator/blob/c4b85072dce3a66eae6dce59e1d7eb20090f8a86/helm/templates/all.yaml), it'd be better if [`hack/helm/pre_helm.go`](https://github.com/TykTechnologies/tyk-operator/blob/cd71d5254803def0a80ff563794f0a3b200fcf74/hack/helm/pre_helm.go) is updated based on your desired changes.

Let me be clear, assume you want to add an option to the helm chart to configure `healthProbeBindAddress` of controller, as you did [here](https://github.com/TykTechnologies/tyk-operator/pull/537/files#diff-88b311698c7c3bad1c1000146aa647d94a4da782a90efc15cb002a328af0903fR357). In order to do that, you can update [`config/manager/controller_manager_config.yaml`](https://github.com/TykTechnologies/tyk-operator/blob/d39e3dd7a9c7f7c6a854dff0ffe0783c9b926378/config/manager/controller_manager_config.yaml#L4)'s `healthProbeBindAddress` to a certain variable, something like `CONTROLLER_MANAGER_HEALTH_PROBE_BINDING_ADDR`. Then, in the [`hack/helm/pre_helm.go`](https://github.com/TykTechnologies/tyk-operator/blob/cd71d5254803def0a80ff563794f0a3b200fcf74/hack/helm/pre_helm.go) script, you can add the following field to the struct
```go
		{"CONTROLLER_MANAGER_HEALTH_PROBE_BINDING_ADDR", healthProbe},
```

to replace occurrences of `CONTROLLER_MANAGER_HEALTH_PROBE_BINDING_ADDR` with `healthProbe`, which can be defined as:
```go
const healthProbe = `:{{ .Values.healthProbePort }}`
```
in the same pre_helm.go file. So that the `kustomize` output will contain 
```yaml
apiVersion: v1
data:
  controller_manager_config.yaml: |
    apiVersion: controller-runtime.sigs.k8s.io/v1alpha1
    kind: ControllerManagerConfig
    health:
      healthProbeBindAddress: CONTROLLER_MANAGER_HEALTH_PROBE_BINDING_ADDR
#rest of the file ...
``` 
and pre_helm.go script will replace all occurrences of `CONTROLLER_MANAGER_HEALTH_PROBE_BINDING_ADDR` with `healthProbe` constant which is defined as above ``:{{ .Values.healthProbePort }}``

I hope I am clear about how to approach the solution. Please let me know if you need further help. I am happy to help. Again, thank you for coming up with this PR 🙏

**Comment by rdcwaldrop1** [comment]
Sorry, for the slow turnaround on this one @buraksekili. I believe I have made the changes you suggested in order to render the `all.yaml` manifest with the `make helm` command. Please let me know if there is anything else I need to add to this PR.  If it looks good I'll mark it as ready for review.  Thanks!

**Comment by rdcwaldrop1** [comment]
Cool, I'll mark as ready for review and try regenerating with a matching version of kustomize.

**Comment by rdcwaldrop1** [comment]
I've bumped my tool versions and still seem to be getting the same formatting issues.  Would it be best to just remove `helm/crds/crds.yaml` from the commit?

**Comment by buraksekili** [comment]
>Would it be best to just remove helm/crds/crds.yaml from the commit? 
i appreciate if you can. sorry for confusion.

File: helm/values.yaml [hunk 1]
```diff
@@ -15,6 +15,9 @@ image:
 imagePullSecrets: []
 nameOverride: ""
 fullnameOverride: ""
+healthProbePort: 8081
+metricsPort: 8080
+webhookPort: 9443
```
Comment (buraksekili) [inline helm/values.yaml:20]: can you please write a small comment about what we are confugiring here? 
For example,
```yaml
# healthProbePort identifies port for ....
healthProbePort: 8081
````

**Comment by buraksekili** [comment]
hi @rdcwaldrop1, do you need some guidance on the PR? I am more than happy to help if you need it.

**Comment by rdcwaldrop1** [comment]
> hi @rdcwaldrop1, do you need some guidance on the PR? I am more than happy to help if you need it.

Hi! In speaking with our solutions engineer he mentioned that this was something that was fixed in the next release. Maybe I misunderstood. If this is still something needed by the project I'm happy to finish this out.

**Comment by buraksekili** [comment]
hi again @rdcwaldrop1 , sorry for the back and forth. We, as a team, want to enable this feature on Tyk Operator. 
Since you have already developed it, It'd be great if you can update the PR and we can go with your contribution. If you have no availability, please let me know. So that I can pick your work and continue working on it. Again, thank you for your contribution 🙏

**Comment by rdcwaldrop1** [comment]
> hi again @rdcwaldrop1 , sorry for the back and forth. We, as a team, want to enable this feature on Tyk Operator. Since you have already developed it, It'd be great if you can update the PR and we can go with your contribution. If you have no availability, please let me know. So that I can pick your work and continue working on it. Again, thank you for your contribution 🙏

Thats fantastic! I can absolutely finish this out. Glad to have the opportunity. I'll make the updates mentioned above today.

**Comment by rdcwaldrop1** [comment]
> > Would it be best to just remove helm/crds/crds.yaml from the commit?
> 
> i appreciate if you can. sorry for confusion.

This file has been removed

File: helm/values.yaml (hunk 1 above)
Comment (rdcwaldrop1) [reply helm/values.yaml]: I've added comments for the port variables

**Review by buraksekili — APPROVED** [review]
LGTM!

**Comment by buraksekili** [comment]
thank you for your contribution @rdcwaldrop1! we really appreciate your time and effort.

Your changes will be available in v0.14.0 :)

**Comment by rdcwaldrop1** [comment]
> thank you for your contribution @rdcwaldrop1! we really appreciate your time and effort.
> 
> Your changes will be available in v0.14.0 :)

Thanks! It was great working with you. I look forward to contributing more in the future.

**Pull request merged**